org.eclipse.scanning.api/src/org/eclipse/scanning/api/scan/event/Location.java

```diff
@@ -1,150 +1,150 @@
-			return false;
-		if (levelNames == null) {
-			if (other.levelNames != null)
-				return false;
-		} else if (!levelNames.equals(other.levelNames))
-			return false;
-		if (name == null) {
-			if (other.name != null)
-				return false;
-		} else if (!name.equals(other.name))
-			return false;
-		if (position == null) {
-			if (other.position != null)
-				return false;
-		} else if (!position.equals(other.position))
-			return false;
-		if (type != other.type)
-			return false;
-		return true;
-	}
-	@Override
-	public String toString() {
-		return "Location [type=" + type + ", name=" + name + ", position=" + position + ", level=" + level + "]";
-	}
-}
+/*-
+ *******************************************************************************
+ * Copyright (c) 2011, 2016 Diamond Light Source Ltd.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Matthew Gerring - initial API and implementation and/or initial documentation
+ *******************************************************************************/
+package org.eclipse.scanning.api.scan.event;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.scanning.api.ILevel;
+import org.eclipse.scanning.api.INameable;
+import org.eclipse.scanning.api.points.IPosition;
+import org.eclipse.scanning.api.scan.PositionEvent;
+
+/**
+ *
+ * A location is a json object for notifying over JMS
+ * the value of an IScannable. We do not use IPosition
+ * directly because this does not have all the information
+ * required. A location may be a transient state which a
+ * motor is moving through or a notifcation of an intent
+ * to move. It can be thought of as a PositionEvent
+ * encapsulated in a single JSON object and with context
+ * information.
+ *
+ * @author Matthew Gerring
+ *
+ */
+public class Location implements INameable {
+
+	/**
+	 * Must match method names in IPositionListener as reflection used on the value to
+	 * call the correct method on the listener.
+	 */
+	public enum LocationType {
+		positionWillPerform, levelPerformed, positionChanged, positionPerformed;
+	}
+
+
+	private LocationType type;
+	private String       name;
+	private IPosition    position;
+	private int          level;
+	private List<String> levelNames;
+
+	public Location() {
+
+	}
+	public Location(LocationType t, PositionEvent evnt) {
+		this.type       = t;
+		this.position   = evnt.getPosition();
+		if (position.getNames()!=null&&position.getNames().size()==1) this.name = position.getNames().iterator().next();
+		this.level      = evnt.getLevel();
+		this.levelNames = readLevelNames(evnt.getLevelObjects());
+	}
+
+	private List<String> readLevelNames(List<? extends ILevel> levelObjects) {
+
+		if (levelObjects==null || levelObjects.isEmpty()) return null;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Return an empty collection instead of null. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1168)